### PR TITLE
[Merged by Bors] - feat: Port positivity extensions for `Nat.ceil`, `Int.ceil`, `Int.floor`

### PR DESCRIPTION
**Pull request by dwrensha**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Order/Floor.lean [hunk 1]
```diff
@@ -1720,61 +1720,58 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
...
+  | .positive pa => pure (.nonnegative (← mkAppM ``int_floor_nonneg_of_pos #[pa]))
+  | .nonnegative pa => pure (.nonnegative (← mkAppM ``int_floor_nonneg #[pa]))
+  | _ => pure .none
+
+private theorem nat_ceil_pos [LinearOrderedSemiring α] [FloorSemiring α] {a : α} :
+    0 < a → 0 < ⌈a⌉₊ :=
+  Nat.ceil_pos.2
+
+/-- Extension for the `positivity` tactic: `Nat.ceil` is positive if its input is. -/
+ @[positivity ⌈ _ ⌉₊]
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Floor.lean:1752]:
```suggestion
@[positivity ⌈ _ ⌉₊]
```

File: Mathlib/Algebra/Order/Floor.lean [hunk 2]
```diff
@@ -1720,61 +1720,58 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
...
+  let pα' : Q(Zero $α') ← synthInstanceQ q(PartialOrder $α')
+  match ← core zα' pα' a with
+  | .positive pa => pure (.positive (← mkAppM ``nat_ceil_pos #[pa]))
+  | _ => pure .none
+
+private theorem int_ceil_pos [LinearOrderedRing α] [FloorRing α] {a : α} : 0 < a → 0 < ⌈a⌉ :=
+  Int.ceil_pos.2
+
+/-- Extension for the `positivity` tactic: `Int.ceil` is positive/nonnegative if its input is. -/
+ @[positivity ⌈ _ ⌉]
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Floor.lean:1766]:
```suggestion
@[positivity ⌈ _ ⌉]
```

File: Mathlib/Algebra/Order/Floor.lean [hunk 3]
```diff
@@ -1720,61 +1720,58 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
...
+private theorem int_floor_nonneg_of_pos [LinearOrderedRing α] [FloorRing α] {a : α}
+    (ha : 0 < a) :
+    0 ≤ ⌊a⌋ :=
+  int_floor_nonneg ha.le
+
+/-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
+@[positivity ⌊ _ ⌋]
+def evalFloor : PositivityExt where eval {_u _α} _zα _pα e := do
+  let (.app (.app (.app (.app _ (α' : Q(Type))) _) _) a) ← whnfR e
+    | throwError "failed to match on Int.floor application"
```
Comment (semorrison) [inline Mathlib/Algebra/Order/Floor.lean:1739]: Hard to tell what this is matching for!

And is it really meant to be constrained to `Type`?

Comment (eric-wieser) [reply Mathlib/Algebra/Order/Floor.lean:1739]: Can we use a `Qq` match here?

Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean]: >  Hard to tell what this is matching for!

I added some comments and included the `.const` head. Hopefully that makes things clearer.


> And is it really meant to be constrained to Type?

Ah, good point. This should be `Type u`. Fixed.

Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean]: > Can we use a Qq match here?

Maybe! But I was unable to figure out how. Even if we do figure it out, I suspect the result will be more complicated that what I have here.

File: Mathlib/Algebra/Order/Floor.lean [hunk 4]
```diff
@@ -1720,61 +1720,61 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
...
+  int_floor_nonneg ha.le
+
+/-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
+@[positivity ⌊ _ ⌋]
+def evalFloor : PositivityExt where eval {u _α} _zα _pα e := do
+  -- match on `@Int.floor α' _ _ a`
+  let (.app (.app (.app (.app (.const ``Int.floor _) (α' : Q(Type $u))) _) _) a) ← whnfR e
+    | throwError "failed to match on Int.floor application"
+  let zα' : Q(Zero $α') ← synthInstanceQ q(Zero $α')
+  let pα' : Q(Zero $α') ← synthInstanceQ q(PartialOrder $α')
```
Comment (eric-wieser) [inline Mathlib/Algebra/Order/Floor.lean:1742]: This type is wrong here

Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean]: fixed

File: Mathlib/Algebra/Order/Floor.lean [hunk 5]
```diff
@@ -1720,61 +1720,61 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
...
+    (ha : 0 < a) :
+    0 ≤ ⌊a⌋ :=
+  int_floor_nonneg ha.le
+
+/-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
+@[positivity ⌊ _ ⌋]
+def evalFloor : PositivityExt where eval {u _α} _zα _pα e := do
+  -- match on `@Int.floor α' _ _ a`
+  let (.app (.app (.app (.app (.const ``Int.floor _) (α' : Q(Type $u))) _) _) a) ← whnfR e
+    | throwError "failed to match on Int.floor application"
```
Comment (eric-wieser) [inline Mathlib/Algebra/Order/Floor.lean:1740]: The `Qq` syntax is something like
```
  have e : Q(ℤ) := ← whnfR e
  let ~q(@Int.floor $α' $inst1 $inst2 $a) := e | throwError "failed to match on Int.floor application"
```
thought I seem to be having trouble making the rest work

File: Mathlib/Algebra/Order/Floor.lean [hunk 6]
```diff
@@ -1720,61 +1720,61 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
...
+  let zα' : Q(Zero $α') ← synthInstanceQ q(Zero $α')
+  let pα' : Q(PartialOrder $α') ← synthInstanceQ q(PartialOrder $α')
+  match ← core zα' pα' a with
+  | .positive pa => pure (.nonnegative (← mkAppM ``int_floor_nonneg_of_pos #[pa]))
+  | .nonnegative pa => pure (.nonnegative (← mkAppM ``int_floor_nonneg #[pa]))
+  | _ => pure .none
```
Comment (eric-wieser) [inline Mathlib/Algebra/Order/Floor.lean:1746]: Here's a full Qq version:
```suggestion
/-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
@[positivity ⌊ _ ⌋]
def evalFloor : PositivityExt where eval {_u _α} _zα _pα e := do
  -- match on `@Int.floor α' _ _ a`
  have ez : Q(ℤ) := ← whnfR e
  let ~q(@Int.floor $α' $i $j $a) := ez | throwError "failed to match on Int.floor application"
  let zα' : Q(Zero $α') ← synthInstanceQ (u := u_1.succ) _
  let pα' : Q(PartialOrder $α') ← synthInstanceQ (u := u_1.succ) _
  assertInstancesCommute
  match ← core zα' pα' a with
  | .positive pa =>
      letI ret : Q(0 ≤ $ez) := q(int_floor_nonneg_of_pos (α := $α') $pa)
      pure (.nonnegative ret)
  | .nonnegative pa =>
      letI ret : Q(0 ≤ $ez) := q(int_floor_nonneg (α := $α') $pa)
      pure (.nonnegative ret)
  | _ => pure .none
```

File: Mathlib/Algebra/Order/Floor.lean [hunk 7]
```diff
@@ -1720,61 +1720,61 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
...
+
+private theorem nat_ceil_pos [LinearOrderedSemiring α] [FloorSemiring α] {a : α} :
+    0 < a → 0 < ⌈a⌉₊ :=
+  Nat.ceil_pos.2
+
+/-- Extension for the `positivity` tactic: `Nat.ceil` is positive if its input is. -/
+@[positivity ⌈ _ ⌉₊]
+def evalNatCeil : PositivityExt where eval {u _α} _zα _pα e := do
+  -- match on `@Nat.ceil α' _ _ a`
+  let (.app (.app (.app (.app (.const ``Nat.ceil _) (α' : Q(Type $u))) _) _) a) ← whnfR e
```
[the remark this answers is not in the document]
Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean]: Good point. I updated it to grab the correct universe from the `Nat.ceil` const.

File: Mathlib/Algebra/Order/Floor.lean (hunk 6 above)
Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean]: Thanks!
I actually prefer what I have, as it's shorter and easier (for me at least) to understand.

I guess that `u_1` gets inserted into the context from the `~q()` match? Is there a way to make that binding explicit?

Comment (eric-wieser) [reply Mathlib/Algebra/Order/Floor.lean]: Yes, the `u_1` is being inserted unhygienically and I can't work out how to name it.

The advantage of my spelling is that it's much less fragile to argument changes to `Int.floor`.

Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean]: > The advantage of my spelling is that it's much less fragile to argument changes to `Int.floor`.

How so? It seems to me that your matching on `~q(@Int.floor $α' $i $j $a)` is essentially the same as my matching on `(.app (.app (.app (.app (.const ``Int.floor [u']) (α' : Q(Type $u'))) _) _) a)`

Comment (eric-wieser) [reply Mathlib/Algebra/Order/Floor.lean]: Here's a shorter version
```suggestion
/-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
@[positivity ⌊ _ ⌋]
def evalFloor : PositivityExt where eval {_u _α} _zα _pα (e : Q(ℤ)) := do
  let ~q(@Int.floor $α' $i $j $a) := e | throwError "failed to match on Int.floor application"
  match ← core q(inferInstance) q(inferInstance) a with
  | .positive pa =>
      letI ret : Q(0 ≤ $e) := q(int_floor_nonneg_of_pos (α := $α') $pa)
      pure (.nonnegative <| ret)
  | .nonnegative pa =>
      letI ret : Q(0 ≤ $e) := q(int_floor_nonneg (α := $α') $pa)
      pure (.nonnegative ret)
  | _ => pure .none
```

Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean]: ah, that's starting to look more reasonable. I'll try that.

Comment (eric-wieser) [reply Mathlib/Algebra/Order/Floor.lean]: > How so? It seems to me that your matching on `~q(@Int.floor $α' $i $j $a)` is essentially the same as my matching on `` (.app (.app (.app (.app (.const ``Int.floor [u']) (α' : Q(Type $u'))) _) _) a) ``

If the syntax changes, my code will become a Qq error at compile time. Yours will become a match failure at tactic use.

**Comment by dwrensha** [comment]
I pushed a version that uses `Qq` for the matching part of `Nat.ceil`. I still have not figured out how to make it work for the proof construction part, but I'm not too worried about that, as the matching part is where the main concerns were in our discussion above.

File: Mathlib/Algebra/Order/Floor.lean [hunk 8]
```diff
@@ -1720,61 +1720,57 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
...
+open Lean.Meta Qq
+
+private theorem int_floor_nonneg [LinearOrderedRing α] [FloorRing α] {a : α} (ha : 0 ≤ a) :
+    0 ≤ ⌊a⌋ :=
+  Int.floor_nonneg.2 ha
+
+private theorem int_floor_nonneg_of_pos [LinearOrderedRing α] [FloorRing α] {a : α}
+    (ha : 0 < a) :
+    0 ≤ ⌊a⌋ :=
+  int_floor_nonneg ha.le
```
Comment (eric-wieser) [inline Mathlib/Algebra/Order/Floor.lean:1733]: I'm curious if there's now any benefit to these; presumably you could inline them in the `q()`s below. Maybe there's a performance benefit this way? cc @semorrison, @gebner.

**Review by eric-wieser — APPROVED** [review]

File: Mathlib/Algebra/Order/Floor.lean (hunk 8 above)
Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean:1733]: I'm curious too! I kept them like this (as they were in mathlib3) for the sake of making the port as direct as possible.

Comment (dwrensha) [reply Mathlib/Algebra/Order/Floor.lean:1733]: Maybe the point is that in mathlib3, inlining was more painful because there was no `q()`.

**Comment by eric-wieser** [comment]
> I still have not figured out how to make it work for the proof construction part

I pushed a change to do this

**Review by YaelDillies — APPROVED** [review]
Analytic number theory really needs those extensions back.

maintainer merge

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**